bool flag
float32 cX
float32 cY
float32 distance